front_end/test/downloads.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as assert from 'assert';
+import { Downloads } from "../src/backend/downloads.js";
+import * as sinon from 'sinon';
+import axios from 'axios';
+import { describe, it } from 'mocha';
+
+async function fakeAxiosGet(URL) {
+    // Use a specific "wrong URL" to mimic a GET request failing. Return -1 as specified in downloads.js
+    const wrongURL = `https://storage.googleapis.com/wrongBucket/wrongFile`;
+    if (URL == wrongURL) {
+        return -1;
+    }
+
+    // If the URL does not match https://storage.googleapis.com/${bucketName}/${fileName}, throw an error.
+    const regexURL = 'https:\/\/storage\.googleapis\.com\/([a-zA-Z]+(-[a-zA-Z]+)+)[0-9]+\/[0-9]+([a-zA-Z]+(\.[a-zA-Z]+)+)';
+    if (!URL.match(regexURL)) {
+        throw new Error('URL is not well defined');
+    }
+}
+
+function fakeDateNow() {
+    return 1;
+}
+
+describe('downloads', () => {
+    let downloads;
+
+    sinon.stub(Date, "now").callsFake(fakeDateNow);
+    sinon.stub(axios, "get").callsFake(fakeAxiosGet);
+
+    beforeEach(() => {
+        downloads = new Downloads();
```

Comment:
does this need to be instantiated before each test? or could it be instantiated once before all the tests?

Reply:
`downloadFile` sets the class attribute `_builtURL`, so that I could test if the URL was being formed correctly. I chose to create a new instance each time since `_builtURL` updating with each test might throw off new tests written later or by someone else. Right now it doesn't actually change the outcome of the tests though - should I instantiate once instead?